src/components/scroll-view/scroll-view.js

```diff
@@ -0,0 +1,130 @@
+import React, { Component, Children } from 'react';
+import PropTypes from 'prop-types';
+
+import styles from './scroll-view.css';
+import List from '../list';
+
+export default class ScrollView extends Component {
+  static propTypes = {
+    length: PropTypes.number,
+    offset: PropTypes.number,
+    itemHeight: PropTypes.number.isRequired,
+    onUpdate: PropTypes.func,
+    children: PropTypes.node.isRequired
+  };
+
+  static defaultProps = {
+    length: 0,
+    offset: 0
+  };
+
+  state = {
+    offset: this.props.offset,
+    visibleItems: 0
+  };
+
+  // update the DOM element's scrollTop position with our initial
+  // page offset and decide the initial visible item count
+  componentDidMount() {
+    let { itemHeight } = this.props;
+    let { offset } = this.state;
+
+    // $list is populated via the ref in the render method below
+    if (this.$list) {
+      this.$list.scrollTop = offset * itemHeight;
+
+      // adjust the amount of visible items on resize
+      window.addEventListener('resize', this.handleListLayout);
+      this.handleListLayout();
+    }
+  }
+
+  componentWillReceiveProps({ offset }) {
+    // new offset that's different from the state
+    if (offset !== this.props.offset && offset !== this.state.offset) {
+      this.setState({ offset });
+    }
+  }
+
+  // clean up our resize listener
+  componentWillUnmount() {
+    window.removeEventListener('resize', this.handleListLayout);
+  }
+
+  // Handles setting the read offset based on the current scroll
+  // position. Also calls `onPage` when a new page has crossed the
+  // threshold of what's considered to be active the view
+  handleScroll = (e) => {
+    let { itemHeight, onUpdate } = this.props;
+
+    let top = e.target.scrollTop;
+    let offset = Math.floor(top / itemHeight);
+
+    // update impagination's readOffset
+    if (this.state.offset !== offset) {
+      this.setState({ offset });
+      onUpdate(offset);
+    }
+  };
+
+  // Handles updating our visible items count based on the list height
+  handleListLayout = () => {
+    let { itemHeight } = this.props;
+
+    if (this.$list) {
+      let listHeight = this.$list.offsetHeight;
+      let visibleItems = Math.ceil(listHeight / itemHeight);
+      this.setState({ visibleItems });
+    }
+  };
+
+  renderChildren() {
+    let { itemHeight, children } = this.props;
+    let { offset, visibleItems } = this.state;
+    let threshold = 5;
+
+    let calcStyle = i => ({
+      height: itemHeight,
+      top: i * itemHeight
+    });
+
+    // calculate the positions before filtering
+    let items = Children.map(children, (child, i) => (
+      <li key={i} className={styles['list-item']} style={calcStyle(i)}>
+        {child}
+      </li>
+    ));
+
+    return items.filter((child, i) => {
```

Comment:
A filter here is still going to have to iterate over each item, right? So if I'm reading this right, then a list with 10,000 items will need to run this filter on every item in the dataset.... on every re-render.

Alternatively, we can map back from `scrollIndex` to a range of items, `[n...m]` and then take that `slice(n,m)` of the dataset state, and then map _those_ records.

Reply:
`slice` definitely makes sense. The logic in `filter` was left over from when I was doing so in the `dataset.state.map()` function.